plugins/uefi-capsule/fu-uefi-common.c

```diff
@@ -171,17 +171,41 @@ fu_uefi_get_framebuffer_size(guint32 *width, guint32 *height, GError **error)
 	return TRUE;
 }
 
-/* return without the ESP dir prepended */
+/**
+ * fu_uefi_get_esp_path_for_os:
+ * @esp_base: The base path of the EFI System Partition (ESP).
+ *
+ * Retrieves the directory structure of the EFI System Partition (ESP) for
+ * the operating system.
+ *
+ * This function constructs and returns the path of the directory to use
+ * within the ESP based on the provided base path.
+ *
+ * Returns: (transfer full): A newly allocated string containing the directory
+ * structure within the ESP for the operating system. The caller is
+ * responsible for freeing the returned string.
+ */
 gchar *
-fu_uefi_get_esp_path_for_os(void)
+fu_uefi_get_esp_path_for_os(const gchar *esp_base)
 {
 #ifndef EFI_OS_DIR
 	const gchar *os_release_id = NULL;
 	const gchar *id_like = NULL;
 	g_autofree gchar *esp_path = NULL;
+	g_autofree gchar *full_path = NULL;
+	g_autofree gchar *systemd_path = NULL;
+	g_autofree gchar *full_systemd_path = NULL;
 	g_autoptr(GError) error_local = NULL;
-	g_autoptr(GHashTable) os_release = fwupd_get_os_release(&error_local);
+	g_autoptr(GHashTable) os_release = NULL;
+
+	/* distro (or user) is using systemd-boot */
+	systemd_path = g_build_filename("EFI", "systemd", NULL);
+	full_systemd_path = g_build_filename(esp_base, systemd_path, NULL);
+	if (g_file_test(full_systemd_path, G_FILE_TEST_IS_DIR))
+		return g_steal_pointer(&systemd_path);
```

Comment:
wouldn't it be a lot simpler if we just returned the fulll UNIX path? I'm not sure why we originally returned without the ESP dir prepended

Reply:
I thought the same and did this initially but there are some functions that use the value without the full path and then it means splitting again.